src/lib.rs

```diff
@@ -74,178 +78,203 @@ where
                 }
                 Ok(Async::Ready(None)) => {
-                    assert!(self.items.is_empty(), "no clock but there are items");
+                    debug_assert!(
+                        self.as_mut().items().is_empty(),
+                        "Inner buffer is empty, but clock is available."
+                    );
                 }
                 Ok(Async::NotReady) => {}
-                Err(e) => {
-                    if self.items.is_empty() {
-                        return Err(Error(Kind::Timer(e)));
-                    } else {
-                        self.err = Some(Error(Kind::Timer(e)));
-                        return self.flush().map_err(|e| Error(Kind::Inner(e)));
+                Err(_e) => {
+                    if !self.as_mut().items().is_empty() {
+                        *self.as_mut().clock() = None;
+                        return Poll::Ready(Some(self.as_mut().take()));
                     }
                 }
             }
-
-            return Ok(Async::NotReady);
+            return Poll::Pending;
         }
     }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let chunk_len = if self.items.is_empty() { 0 } else { 1 };
+        let (lower, upper) = self.stream.size_hint();
+        let lower = lower.saturating_add(chunk_len);
+        let upper = match upper {
+            Some(x) => x.checked_add(chunk_len),
+            None => None,
+        };
+        (lower, upper)
+    }
+}
+
+impl<St: FusedStream> FusedStream for ChunksTimeout<St> {
+    fn is_terminated(&self) -> bool {
+        self.stream.is_terminated() & self.items.is_empty()
+    }
+}
+
+// Forwarding impl of Sink from the underlying stream
+#[cfg(feature = "sink")]
+impl<S, Item> Sink<Item> for ChunksTimeout<S>
+where
+    S: Stream + Sink<Item>,
+{
+    type Error = S::Error;
+
+    delegate_sink!(stream, Item);
 }
 
 #[cfg(test)]
 mod tests {
     use super::*;
-    use futures::stream;
-    use std::io;
+    use futures::compat::Future01CompatExt;
+    use futures::future;
+    use futures::{stream, FutureExt, StreamExt, TryFutureExt};
     use std::iter;
     use std::time::{Duration, Instant};
 
     #[test]
     fn messages_pass_through() {
-        let iter = iter::once(5);
-        let stream = stream::iter_ok::<_, io::Error>(iter);
-
-        let chunk_stream = Chunks::new(stream, 5, Duration::new(10, 0));
-
-        let v = chunk_stream.collect();
-        tokio::run(v.then(|res| {
-            match res {
-                Err(_) => assert!(false),
-                Ok(v) => assert_eq!(vec![vec![5]], v),
-            };
-            Ok(())
-        }));
+        let v = stream::iter(iter::once(5))
+            .chunks_timeout(5, Duration::new(1, 0))
+            .collect::<Vec<_>>();
+        tokio::run(
+            v.then(|x| {
+                assert_eq!(vec![vec![5]], x);
+                future::ready(())
+            })
+            .unit_error()
+            .boxed()
+            .compat(),
```

Comment:
`.unit_error().boxed().compat()` is quite a lot of bookkeeping. Is there a better way? Probably not.

Reply:
Maybe there's something in `tokio-test` or `futures-test-preview` (that you mentioned below) which could help us here?

Reply:
> `.unit_error().boxed().compat()` is quite a lot of bookkeeping. Is there a better way? Probably not.

Unfortunately not, it's quite verbose.

> Maybe there's something in `tokio-test` or `futures-test-preview` (that you mentioned below) which could help us here?

I think with `futures-test-preview` we can directly drive the streams w/o any externally provided runtime. I might give it a shot.

Reply:
I won't be able to refactor the tests in the near future, so I'd leave the TODO in if you're ok with that.

Reply:
Sure